File: judge/src/tested/languages/java/templates/Values.java
import java.io.PrintWriter;
import java.io.StringWriter;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.*;

/**
 * Minimal RPC language in JSON to send data from the tests to the judge.
 */
public class Values {

    private static String encodeSequence(Iterable<Object> objects) {
        var results = new ArrayList<String>();
        for (Object obj: objects) {
            results.add(encode(obj));
        }
        return "[" + String.join(", ", results) + "]";
    }

    @SuppressWarnings("unchecked")
    private static List<String> internalEncode(Object value) {
        String type;
        String data;

        if (value == null) {
            type = "nothing";
            data = "null";
        } else if (value instanceof Boolean) {
            type = "boolean";
            data = value.toString();
        } else if (value.getClass().isArray()) {
            type = "array";
            data = encodeSequence(Arrays.asList((Object[]) value));
        } else if (value instanceof BigInteger) {
            type = "bigint";
            data = value.toString();
        } else if (value instanceof BigDecimal) {
            type = "fixed_precision";
            data = value.toString();
        } else if (value instanceof Byte) {
            type = "int8";
            data = value.toString();
        } else if (value instanceof Short) {
            type = "int16";
            data = value.toString();
        } else if (value instanceof Integer) {
            type = "int32";
            data = value.toString();
        } else if (value instanceof Long) {
            type = "int64";
            data = value.toString();
        } else if (value instanceof Float) {
            type = "single_precision";
            data = value.toString();
        } else if (value instanceof Double) {
            type = "double_precision";
            data = value.toString();
        } else if (value instanceof Character) {
            type = "char";
            data = "\"" + value.toString() + "\"";
        } else if (value instanceof CharSequence) {
            type = "text";
            data = "\"" + value.toString() + "\"";
        } else if (value instanceof List) {
            type = "list";
            data = encodeSequence((Iterable<Object>) value);
        } else if (value instanceof Set) {
            type = "set";
            data = encodeSequence((Iterable<Object>) value);
        } else if (value instanceof Map) {
            type = "object";
            var elements = new ArrayList<String>();
            for (Map.Entry<Object, Object> entry : ((Map<Object, Object>) value).entrySet()) {
                elements.add("\"" + entry.getKey().toString() + "\": " + encode(entry.getValue()));
            }
            data = "{" + String.join(", ", elements) + "}";
        } else {
            type = "unknown";
            data = value.toString();
        }
        return List.of(type, data);
    }

    private static String encode(Object value) {
        var typeAndData = internalEncode(value);
        return "{ \"data\": " + typeAndData.get(1) + ", \"type\": \"" + typeAndData.get(0) + "\"}";
    }

    public static void send(PrintWriter writer, Object value) {
        writer.print(encode(value));
    }

    public static void sendException(PrintWriter writer, Exception exception) {
        var sw = new StringWriter();
        exception.printStackTrace(new PrintWriter(sw));
        var result = "{ \"message\": \"" + exception.getMessage() + "\", \"type\": \"" + sw.toString() + "\"}";
        writer.write(result);
    }

    public static void evaluated(PrintWriter writer,
                                 boolean result, String expected, String actual, Collection<String> messages) {
        var encodedExpected = internalEncode(expected);
        var encodedActual = internalEncode(actual);
        String builder = "{" +
            "\"result\": " +
            result +
            ", \"readable_expected\": " +
            encodedExpected.get(1) +
            ", \"readable_actual\": " +
            encodedActual.get(1) +
            ", \"messages\": [" +
            String.join(", ", messages) +
            "]}";
        writer.print(builder);
    }
}

File: judge/src/tested/languages/kotlin/templates/Values.kt
import java.io.PrintWriter
import java.io.StringWriter
import java.math.BigDecimal
import java.math.BigInteger

private fun convertMessage(message: EvaluationResult.Message): String {
    val builder = StringBuilder(64).append('{')
            .append("\"description\": \"").append(message.description)
            .append("\", ")
            .append("\"format\": \"").append(message.format).append("\"")
    if (message.permission != null) {
        builder.append(", \"permission\": \"").append(message.permission)
                .append('\"')
    }
    return builder.append('}').toString()
}

private fun escape(str: String): String {
    return str
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
}

private fun encodeSequence(objects: Iterable<Any?>): String {
    return objects.asSequence()
            .joinToString(
                    separator = ", ",
                    prefix = "[",
                    postfix = "]",
                    transform = { o -> encode(o) }
            )
}

private fun encode(value: Any?): String {
    val typeAndData = internalEncode(value)
    return String.format("{ \"data\": %s, \"type\": \"%s\"}",
            typeAndData[1], typeAndData[0])
}

private fun internalEncode(value: Any?): Array<String> {
    val type: String
    val data: String

    if (value == null) {
        type = "nothing"
        data = "null"
    } else if (value is Boolean) {
        type = "boolean"
        data = value.toString()
    } else if (value.javaClass.isArray) {
        type = "array"
        data = encodeSequence((value as Array<*>).asIterable())
    } else if (value is BigInteger) {
        type = "bigint"
        data = value.toString()
    } else if (value is BigDecimal) {
        type = "fixed_precision"
        data = value.toString()
    } else if (value is Byte) {
        type = "int8"
        data = value.toString()
    } else if (value is Short) {
        type = "int16"
        data = value.toString()
    } else if (value is Int) {
        type = "int32"
        data = value.toString()
    } else if (value is Long) {
        type = "int64"
        data = value.toString()
    } else if (value is Char) {
        type = "character"
        data = String.format("\"%s\"", escape(value.toString()))
    } else if (value is CharSequence) {
        type = "text"
        data = String.format("\"%s\"", escape(value.toString()))
    } else if (value is List<*>) {
        type = "list"
        data = encodeSequence(value.asIterable())
    } else if (value is Set<*>) {
        type = "set"
        data = encodeSequence(value.asIterable())
    } else if (value is Map<*, *>) {
        type = "map"
        data = value.asSequence()
                .map { e ->
                    String.format("\"%s\": \"%s\"", e.key.toString(),
                            encode(e.value))
                }
                .joinToString(separator = ", ", prefix = "{", postfix = "}")
    } else {
        type = "unknown"
        data = value.toString()
    }

    return arrayOf(type, data)
}

fun evaluated(writer: PrintWriter, result: Boolean, expected: String?,
              actual: String?, messages: Iterable<EvaluationResult.Message>) {
    val builder = StringBuilder(64).append('{')
            .append("\"result\": \"").append(result).append("\", ")
            .append("\"readable_expected\": \"").append(expected).append("\", ")
            .append("\"readable_actual\": \"").append(actual).append("\", ")
            .append("\"messages\": [").append(messages.joinToString(separator = ", ",
                    transform = { m -> convertMessage(m) })).append("]}")
    writer.print(builder.toString())
}

fun valuesSend(writer: PrintWriter, value: Any?): Unit = writer.print(encode(value))

fun valuesSendEvaluated(writer: PrintWriter, result: EvaluationResult) {
    evaluated(writer, result.result, result.readableExpected, result.readableActual,
            result.messages)
}

fun valuesSendException(writer: PrintWriter, throwable: Throwable?) {
    if (throwable == null) {
        return
    }
    val strStackTraceWriter = StringWriter()
    throwable.printStackTrace(PrintWriter(strStackTraceWriter))
    writer.printf("{ \"message\": \"%s\", \"type\": \"%s\"}",
            throwable.message, strStackTraceWriter.toString())
}
